File: job_selection.c
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
//
// AED, 2020/2021
//
// TODO: Pedro Sobral, 98491
// TODO: André Freixo, 98495
// TODO: Marta Fradique, 98626
//
// Brute-force solution of the generalized weighted job selection problem
//
// Compile with "cc -Wall -O2 job_selection.c -lm" or equivalent
//
// In the generalized weighted job selection problem we will solve here we have T programming tasks and P programmers.
// Each programming task has a starting date (an integer), an ending date (another integer), and a profit (yet another
// integer). Each programming task can be either left undone or it can be done by a single programmer. At any given
// date each programmer can be either idle or it can be working on a programming task. The goal is to select the
// programming tasks that generate the largest profit.
//
// Things to do:
//   0. (mandatory)
//      Place the student numbers and names at the top of this file.
//   1. (highly recommended)
//      Read and understand this code.
//   2. (mandatory)
//      Solve the problem for each student number of the group and for
//        N=1, 2, ..., as higher as you can get and
//        P=1, 2, ... min(8,N)
//      Present the best profits in a table (one table per student number).
//      Present all execution times in a graph (use a different color for the times of each student number).
//      Draw the solutions for the highest N you were able to do.
//   3. (optional)
//      Ignore the profits (or, what is the same, make all profits equal); what is the largest number of programming
//      tasks that can be done?
//   4. (optional)
//      Count the number of valid task assignments. Calculate and display an histogram of the number of occurrences of
//      each total profit. Does it follow approximately a normal distribution?
//   5. (optional)
//      Try to improve the execution time of the program (use the branch-and-bound technique).
//      Can you use divide and conquer to solve this problem?
//      Can you use dynamic programming to solve this problem?
//   6. (optional)
//      For each problem size, and each student number of the group, generate one million (or more!) valid random
//      assignments and compute the best solution found in this way. Compare these solutions with the ones found in
//      item 2.
//   7. (optional)
//      Surprise us, by doing something more!
//   8. (mandatory)
//      Write a report explaining what you did. Do not forget to put all your code in an appendix.
//

#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <sys/types.h>
#include "../P02/elapsed_time.h"


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
//
// Random number generator interface (do not change anything in this code section)
//
// In order to ensure reproducible results on Windows and GNU/Linux, we use a good random number generator, available at
//   https://www-cs-faculty.stanford.edu/~knuth/programs/rng.c
// This file has to be used without any modifications, so we take care of the main function that is there by applying
// some C preprocessor tricks
//

#define main  rng_main                        // main gets replaced by rng_main
#ifdef __GNUC__
int rng_main() __attribute__((__unused__));   // gcc will not complain if rnd_main() is not used
#endif
#include "rng.c"
#undef main                                   // main becomes main again

#define srandom(seed)  ran_start((long)seed)  // start the pseudo-random number generator
#define random()       ran_arr_next()         // get the next pseudo-random number (0 to 2^30-1)


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
//
// problem data (if necessary, add new data fields in the structures; do not change anything else in this code section)
//
// on the data structures declared below, a comment starting with
// * a I means that the corresponding field is initialized by init_problem()
// * a S means that the corresponding field should be used when trying all possible cases
// * IS means both (part initialized, part used)
//

#if 1

#define MAX_T  64  // maximum number of programming tasks
#define MAX_P  10  // maximum number of programmers

typedef struct
{
  int starting_date;      // I starting date of this task
  int ending_date;        // I ending date of this task
  int profit;             // I the profit if this task is performed
  int assigned_to;        // S current programmer number this task is assigned to (use -1 for no assignment)
}
task_t;

typedef struct
{
  int NMec;               // I  student number
  int T;                  // I  number of tasks
  int P;                  // I  number of programmers
  int I;                  // I  if 1, ignore profits
  int total_profit;       // S  current total profit
  double cpu_time;        // S  time it took to find the solution
  task_t task[MAX_T];     // IS task data
  int busy[MAX_P];        // S  for each programmer, record until when she/he is busy (-1 means idle)
  char dir_name[16];      // I  directory name where the solution file will be created
  char file_name[64];     // I  file name where the solution data will be stored
}
problem_t;

int compare_tasks(const void *t1,const void *t2)
{
  int d1,d2;

  d1 = ((task_t *)t1)->starting_date;
  d2 = ((task_t *)t2)->starting_date;
  if(d1 != d2)
    return (d1 < d2) ? -1 : +1;
  d1 = ((task_t *)t1)->ending_date;
  d2 = ((task_t *)t2)->ending_date;
  if(d1 != d2)
    return (d1 < d2) ? -1 : +1;
  return 0;
}

void init_problem(int NMec,int T,int P,int ignore_profit,problem_t *problem)
{
  int i,r,scale,span,total_span;
  int *weight;

  //
  // input validation
  //

  if(NMec < 1 || NMec > 999999)
  {
    fprintf(stderr,"Bad NMec (1 <= NMex (%d) <= 999999)\n",NMec);
    exit(1);
  }
  if(T < 1 || T > MAX_T)
  {
    fprintf(stderr,"Bad T (1 <= T (%d) <= %d)\n",T,MAX_T);
    exit(1);
  }
  if(P < 1 || P > MAX_P)
  {
    fprintf(stderr,"Bad P (1 <= P (%d) <= %d)\n",P,MAX_P);
    exit(1);
  }
  //
  // the starting and ending dates of each task satisfy 0 <= starting_date <= ending_date <= total_span
  //
  total_span = (10 * T + P - 1) / P;
  if(total_span < 30)
    total_span = 30;
  //
  // probability of each possible task duration
  //
  // task span relative probabilities
  //
  // |  0  0  4  6  8 10 12 14 16 18 | 20 | 19 18 17 16 15 14 13 12 11 10  9  8  7  6  5  4  3  2  1 | smaller than 1
  // |  0  0  2  3  4  5  6  7  8  9 | 10 | 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 | 30 31 ... span
  //
  weight = (int *)alloca((size_t)(total_span + 1) * sizeof(int)); // allocate memory (freed automatically)
  if(weight == NULL)
  {
    fprintf(stderr,"Strange! Unable to allocate memory\n");
    exit(1);
  }
#define sum1  (298.0)                      // sum of weight[i] for i=2,...,29 using the data given in the comment above
#define sum2  ((double)(total_span - 29))  // sum of weight[i] for i=30,...,data_span using a weight of 1
#define tail  100
  scale = (int)ceil((double)tail * 10.0 * sum2 / sum1); // we want that scale*sum1 >= 10*tail*sum2, so that large task
  if(scale < tail)                                      // durations occur 10% of the time
    scale = tail;
  weight[0] = 0;
  weight[1] = 0;
  for(i = 2;i <= 10;i++)
    weight[i] = scale * (2 * i);
  for(i = 11;i <= 29;i++)
    weight[i] = scale * (30 - i);
  for(i = 30;i <= total_span;i++)
    weight[i] = tail;
#undef sum1
#undef sum2
#undef tail
  //
  // accumulate the weigths (cummulative distribution)
  //
  for(i = 1;i <= total_span;i++)
    weight[i] += weight[i - 1];
  //
  // generate the random tasks
  //

  srandom(NMec + 314161 * T + 271829 * P);
  problem->NMec = NMec;
  problem->T = T;
  problem->P = P;
  problem->I = (ignore_profit == 0) ? 0 : 1;
  for(i = 0;i < T;i++)
  {
    //
    // task starting an ending dates
    //
    r = 1 + (int)random() % weight[total_span]; // 1 .. weight[total_span]
    for(span = 0;span < total_span;span++)
      if(r <= weight[span])
        break;
    problem->task[i].starting_date = (int)random() % (total_span - span + 1); 
    problem->task[i].ending_date = problem->task[i].starting_date + span - 1;
    //
    // task profit
    //
    // the task profit is given by r*task_span, where r is a random variable in the range 50..300 with a probability
    //   density function with shape (two triangles, the area of the second is 4 times the area of the first)
    //
    //       *
    //*     /|   *
    // *   / |       *
    //*   /  |           *
    //   *---*---------------*
    //  50 100 150 200 250 300
    //
    scale = (int)random() % 12501; // almost uniformly distributed in 0..12500
    if(scale <= 2500)
      problem->task[i].profit = 1 + round((double)span * (50.0 + sqrt((double)scale)));
    else
      problem->task[i].profit = 1 + round((double)span * (300.0 - 2.0 * sqrt((double)(12500 - scale))));
  }
  //
  // sort the tasks by the starting date
  //
  qsort((void *)&problem->task[0],(size_t)problem->T,sizeof(problem->task[0]),compare_tasks);
  //
  // finish
  //
  if(problem->I != 0)
    for(i = 0;i < problem->T;i++)
      problem->task[i].profit = 1;
#define DIR_NAME  problem->dir_name
  if(snprintf(DIR_NAME,sizeof(DIR_NAME),"%06d",NMec) >= sizeof(DIR_NAME))
  {
    fprintf(stderr,"Directory name too large!\n");
    exit(1);
  }
#undef DIR_NAME
#define FILE_NAME  problem->file_name
  if(snprintf(FILE_NAME,sizeof(FILE_NAME),"%06d/%02d_%02d_%d.txt",NMec,T,P,problem->I) >= sizeof(FILE_NAME))
  {
    fprintf(stderr,"File name too large!\n");
    exit(1);
  }
#undef FILE_NAME
}

#endif


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
//
// problem solution (place your solution here)
//

#if 1

static void solve(problem_t *problem)
{
  FILE *fp;
  int i;

  //
  // open log file
  //
  (void)mkdir(problem->dir_name,S_IRUSR | S_IWUSR | S_IXUSR);
  fp = fopen(problem->file_name,"w");
  if(fp == NULL)
  {
    fprintf(stderr,"Unable to create file %s (maybe it already exists? If so, delete it!)\n",problem->file_name);
    exit(1);
  }
  //
  // solve
  //
  problem->cpu_time = cpu_time();
    // call your (recursive?) function to solve the problem here
  problem->cpu_time = cpu_time() - problem->cpu_time;
  //
  // save solution data
  //
  fprintf(fp,"NMec = %d\n",problem->NMec);
  fprintf(fp,"T = %d\n",problem->T);
  fprintf(fp,"P = %d\n",problem->P);
  fprintf(fp,"Profits%s ignored\n",(problem->I == 0) ? " not" : "");
  fprintf(fp,"Solution time = %.3e\n",problem->cpu_time);
  fprintf(fp,"Task data\n");
#define TASK  problem->task[i]
  for(i = 0;i < problem->T;i++)
    fprintf(fp,"  %3d %3d %5d\n",TASK.starting_date,TASK.ending_date,TASK.profit);
#undef TASK
  fprintf(fp,"End\n");
  //
  // terminate
  //
  if(fflush(fp) != 0 || ferror(fp) != 0 || fclose(fp) != 0)
  {
    fprintf(stderr,"Error while writing data to file %s\n",problem->file_name);
    exit(1);
  }
}

#endif



////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
//
// main program
//

int main(int argc,char **argv)
{
  problem_t problem;
  int NMec,T,P,I;

  NMec = (argc < 2) ? 2020 : atoi(argv[1]);
  T = (argc < 3) ? 5 : atoi(argv[2]);
  P = (argc < 4) ? 2 : atoi(argv[3]);
  I = (argc < 5) ? 0 : atoi(argv[4]);
  init_problem(NMec,T,P,I,&problem);
  solve(&problem);
  return 0;
}

File: Trabalho01_AED_JAVA/src/TRABALHO01_AED_JAVA/JobSelection.java
package TRABALHO01_AED_JAVA;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

import javax.print.attribute.standard.PrinterInfo;

public class JobSelection {
    public static void main(String[] args) {

        Problem_t problem =  null;
        int NMEC, T, P, I;

        NMEC = Integer.parseInt(args[0]);
        T = Integer.parseInt(args[1]);
        P = Integer.parseInt(args[2]);
        I = Integer.parseInt(args[3]);
        init_problem(NMEC, T, P, I, problem);
        solve(problem);
    }

    public int compare_tasks(Task_t t1, Task_t t2) {
        int d1, d2;

        d1 = t1.getStarting_date();
        d2 = t2.getStarting_date();

        if (d1 != d2) {
            return (d1 < d2) ? -1 : +1;
        }

        d1 = t1.getEnding_date();
        d2 = t2.getEnding_date();
        if (d1 != d2) {
            return (d1 < d2) ? -1 : +1;
        }
        return 0;
    }

    public static void init_problem(int NMec, int T, int P, int ignore_profit, Problem_t problem) {

        int i, r, scale, span, total_span;
        List<Integer> weight = new ArrayList<>();

        // Input validation
        if (NMec < 1 || NMec > 999999) {
            System.out.printf("Bad NMEC (1 <= NMec (%d)) <= 999999", NMec);
            return;
        }
        if (T < 1 || T > 64) {
            System.out.printf("Bad T (1 <= T (%d) <= 64)", T);
            return;
        }
        if (P < 1 || P > 10) {
            System.out.printf("Bad P (1 <= P (%d) <= %d)", P);
            return;
        }

        // the starting and ending dates of each task satisfy 0 <= starting_date <=
        // ending_date <= total_span
        total_span = (10 * T + P - 1) / P;
        if (total_span < 30) {
            total_span = 30;
        }

        // probability of each possible task duration
        // ! Nao sei como se faz para alocar memoria em Java, ou qual a operacao
        // semelhante, ou se nem há
        // weight = (int *)alloca((size_t)(total_span + 1) * sizeof(int)); // allocate
        // memory (freed automatically)
        // if (weight == NULL) {
        // fprintf(stderr, "Strange! Unable to allocate memory\n");
        // exit(1);
        // }

        int tail = 1; // ! NAO SEI O QUE É O TAIL

        scale = (int) Math.ceil(tail * 10.0 * ((double) total_span - 29) / 298.0);
        if (scale < tail) {
            scale = tail;
        }
        // weight[0] = 0;
        // weight[1] = 0;

        for (i = 2; i <= 10; i++) {
            weight.add(i, scale * (2 * i));
        }
        for (i = 11; i <= 29; i++) {
            weight.add(i, scale * (30 - i));
        }
        for (i = 30; i <= total_span; i++) {
            weight.add(i, tail);
        }

        // accumulate the weigths (cummulative distribution)
        int sum = 0;
        for (i = 1; i <= total_span; i++) {
            sum += weight.get(i - 1);
            weight.add(i, sum);
        }

        // ! numero random

        problem.setnMec(NMec);
        problem.setT(T);
        problem.setP(P);
        problem.setnMec((ignore_profit == 0) ? 0 : 1);

        for (i = 0; i < T; i++) {
            // tasks starting and ending dates
            r = 1 + (int) Math.random() % weight.get(total_span);

            for (span = 0; span < total_span; span++) {
                if (r <= weight.get(span)) {
                    break;
                }

                problem.task.get(i).setStarting_date((int) Math.random() % (total_span - span + 1));
                problem.task.get(i).setEnding_date(problem.task.get(i).getStarting_date() + span - 1);

            }

            scale = (int) Math.random() % 12501;
            if (scale <= 2500) {
                problem.task.get(i).setProfit(1 + Math.round((double) span) * (50.0 + Math.sqrt((double) scale)));
            } else {
                problem.task.get(i).setProfit(1 + Math.round((double) span) * (300.0 - 2.0 * Math.sqrt((double) (12500 - scale))));
            }

            // sort the tasks by the starting date
        }
        // finish

        if (problem.getI() != 0) {
            for (i = 0; i < problem.getT(); i++) {
                problem.task.get(i).getProfit() = 1;
            }
        }
    }

    public void recursive_function(Problem_t problem, int i) {

        if (i == 0) {
            problem.setTotal_profit(0);
            problem.setBest_total_profit(0);
            for (int k = 0; k < problem.getP(); k++) {
                problem.busy.get(k).add(-1);
            }
        }

        if (i == problem.getT()) {
            if (problem.getTotal_profit() > problem.getBest_total_profit()) {
                problem.setBest_total_profit(problem.getTotal_profit());
            }
            return;
        }

        problem.task.get(i).setAssigned_to(-1);

        recursive_function(problem, i + 1);

        for (int j = 0; j < problem.getP(); j++) {
            if (problem.busy.get(j) < problem.task.get(i).getStarting_date()) {
                int profit_tmp = problem.getTotal_profit();
                int busy_tmp = problem.busy.get(j);

                problem.busy.get(j).setBusy(problem.task.get(i).getEnding_date());
                problem.setTotal_profit((problem.getTotal_profit() + problem.task.get(i).getProfit()));
                problem.task.get(i).setAssigned_to(-1);

                recursive_function(problem, i + 1);

                problem.task.get(i).setAssigned_to(-1);
                problem.setTotal_profit(profit_tmp);
                problem.busy.get(j) = busy_tmp;

                return;
            }
        }
    }

    public static void solve(Problem_t problem) {

        StringBuilder nameFile = new StringBuilder();
        nameFile.append(problem.getT() + "_" + problem.getP() + "_" + problem.getI() + ".txt");

        try (PrintWriter out = new PrintWriter(new File( nameFile ))) {
            out.printf("NMec = %d\n", problem.getnMec());
            out.printf("T = %d\n", problem.getT());
            out.printf("P = %d\n", problem.getP());
            out.printf("Profits%s  = %d\n", (problem.getI() == 0) ? " not" : "");
            out.printf("Best Profit = %d\n", problem.getBest_total_profit());
            out.printf("Best Assigned = %d\n", problem.getBest_assigned_to());
            out.printf("Solution time = %.3e\n", problem.getCpu_time());
            out.printf("Task date  Profit    P\n");

            for (int i = 0; i < problem.getT(); i++) {
                out.printf("%-3d %-3d   %-8d %-4d\n", problem.task.get(i).getStarting_date(),
                        problem.task.get(i).getEnding_date(), problem.task.get(i).getProfit(),
                        problem.task.get(i).getAssigned_to());
            }
            out.printf("End\n");

        } catch (FileNotFoundException e) {
            System.err.println(e.getMessage());
        }
    }
}
